File: StudyNotes/GeekBang.org/视频课/算法面试通关40讲/08 | 面试题：判断括号字符串是否有效/0020/20.go
func isValid(s string) bool {
	stack := []byte{}
	paren_map := map[byte]byte{')': '(', ']': '[', '}': '{'}
	for i := 0; i < len(s); i++ {
		c := s[i]
		if _, ok := paren_map[c]; !ok {
			stack = append(stack, c)
		} else if len(stack) == 0 || paren_map[c] != stack[len(stack)-1] {
			return false
		} else {
      stack = stack[:len(stack)-1]
    }
	}
	return len(stack) == 0
}

File: StudyNotes/GeekBang.org/视频课/算法面试通关40讲/08 | 面试题：判断括号字符串是否有效/0020/20.js
var isValid = function (s) {
  const stack = [];
  const paren_map = new Map([
    [")", "("],
    ["]", "["],
    ["}", "{"],
  ]);
  for (let c of s) {
    if (!paren_map.has(c)) {
      stack.push(c);
    } else if (!stack.length || paren_map.get(c) !== stack.pop()) {
      return false;
    }
  }
  return !stack.length;
};
